File: crates/gdext/src/commands/script_helpers.rs
use serde_json::{Value, json};

use godot::prelude::StringName;

use super::{get_root, j2v, pipe, resolve_node, s, v2j};
use crate::dispatcher::MainThreadDispatcher;

pub const TOOL_NAMES: &[&str] = &["call_method", "get_variable", "set_variable"];

pub struct ScriptHelpersCommands;

impl ScriptHelpersCommands {
    pub fn new() -> Self {
        Self
    }
}

impl Default for ScriptHelpersCommands {
    fn default() -> Self {
        Self::new()
    }
}

impl super::CommandHandler for ScriptHelpersCommands {
    fn can_handle(&self, tool: &str) -> bool {
        TOOL_NAMES.contains(&tool)
    }
    fn handle<'a>(
        &'a self,
        tool: &'a str,
        args: &'a Value,
        d: &'a MainThreadDispatcher,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Value, String>> + Send + 'a>>
    {
        Box::pin(self.handle_script_helpers_tool(tool, args, d))
    }
    fn group_name(&self) -> &str {
        "script_helpers"
    }
    fn tool_names(&self) -> &[&str] {
        TOOL_NAMES
    }
}

impl ScriptHelpersCommands {
    pub async fn handle_script_helpers_tool(
        &self,
        tool: &str,
        args: &Value,
        d: &MainThreadDispatcher,
    ) -> Result<Value, String> {
        let a = args.clone();
        match tool {
            "call_method" => pipe(d.submit(move || cmd_call_method(&a)).await),
            "get_variable" => pipe(d.submit(move || cmd_get_variable(&a)).await),
            "set_variable" => pipe(d.submit(move || cmd_set_variable(&a)).await),
            _ => Err(format!("Unknown script_helpers tool: {}", tool)),
        }
    }
}

fn cmd_call_method(args: &Value) -> Value {
    let p = s(args, "node_path");
    let method = s(args, "method");
    if method.is_empty() {
        return json!({"error": "missing 'method'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    let mut node = match resolve_node(&root, p.as_str()) {
        Some(n) => n,
        None => return json!({"error": format!("Node not found: {}", p)}),
    };
    let call_args: Vec<godot::prelude::Variant> = args["args"]
        .as_array()
        .map(|arr| arr.iter().map(j2v).collect())
        .unwrap_or_default();
    let result = node.call(&StringName::from(method.as_str()), &call_args);
    let is_nil = result.is_nil();
    let mut resp = json!({
        "node_path": p,
        "method": method,
        "result": v2j(&result),
        "type": format!("{:?}", result.get_type())
    });
    if is_nil {
        resp.as_object_mut().unwrap().insert(
            "hint".into(),
            json!("Method returned null. In editor mode, custom GDScript methods can only execute on @tool scripts. \
                   Add '@tool' at the top of the script and re-attach it to enable editor-time method calls. \
                   Built-in engine methods (e.g. get_name, get_child_count) work regardless of @tool."),
        );
    }
    resp
}

fn cmd_get_variable(args: &Value) -> Value {
    let p = s(args, "node_path");
    let var_name = s(args, "variable");
    if var_name.is_empty() {
        return json!({"error": "missing 'variable'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    match resolve_node(&root, p.as_str()) {
        Some(n) => {
            let val = n.get(&StringName::from(var_name.as_str()));
            if val.is_nil() {
                let retry = n.get(&StringName::from(var_name.as_str()));
                if !retry.is_nil() {
                    return json!({
                        "node_path": p,
                        "variable": var_name,
                        "value": v2j(&retry),
                        "type": format!("{:?}", retry.get_type())
                    });
                }
                json!({
                    "node_path": p,
                    "variable": var_name,
                    "value": null,
                    "type": "NIL",
                    "hint": format!(
                        "Variable '{}' not found. In editor mode, Object.get() only returns \
                         engine properties and @export script variables. \
                         Use @export var {} in your GDScript to make it accessible.",
                        var_name, var_name
                    )
                })
            } else {
                json!({
                    "node_path": p,
                    "variable": var_name,
                    "value": v2j(&val),
                    "type": format!("{:?}", val.get_type())
                })
            }
        }
        None => json!({"error": format!("Node not found: {}", p)}),
    }
}

fn cmd_set_variable(args: &Value) -> Value {
    let p = s(args, "node_path");
    let var_name = s(args, "variable");
    let val = match args.get("value") {
        Some(v) => v.clone(),
        None => return json!({"error": "missing 'value'"}),
    };
    if var_name.is_empty() {
        return json!({"error": "missing 'variable'"});
    }
    let root = match get_root() {
        Ok(r) => r,
        Err(e) => return e,
    };
    match resolve_node(&root, p.as_str()) {
        Some(mut n) => {
            let gv = j2v(&val);
            n.set(&StringName::from(var_name.as_str()), &gv);
            let actual = n.get(&StringName::from(var_name.as_str()));
            if actual.is_nil() {
                return json!({
                    "node_path": p,
                    "variable": var_name,
                    "success": false,
                    "hint": format!(
                        "Variable '{}' was not actually set. In editor mode, Object.set() only works \
                         with engine properties and @export script variables. Non-@export variables \
                         are stored in PlaceHolderScriptInstance which may not persist the value. \
                         Use @export var {} in your GDScript.",
                        var_name, var_name
                    )
                });
            }
            json!({
                "node_path": p,
                "variable": var_name,
                "value": v2j(&actual),
                "success": true
            })
        }
        None => json!({"error": format!("Node not found: {}", p)}),
    }
}

File: extensions/gdext/src/commands/script_helpers.cpp
// commands/script_helpers.cpp — call_method, get_variable, set_variable

#include "cmd_utils.hpp"
#include "handler_registry.hpp"
#include <godot_cpp/classes/script.hpp>
#include <godot_cpp/variant/array.hpp>
using namespace godot;

namespace godot_mcp {
namespace {

Dictionary cmd_call_method(const Dictionary &a) {
    String p = args_string(a, "node_path");
    String method = args_string(a, "method");
    if (method.is_empty()) return make_error("missing 'method'");
    Node *root = get_root(); if (!root) return make_error("no scene");
    Node *n = resolve_node(root, p); if (!n) return make_error("node not found: " + p);
    Array call_args;
    if (a.has("args") && a["args"].get_type() == Variant::ARRAY) {
        Array raw = a["args"];
        for (int i = 0; i < raw.size(); i++) call_args.append(json_to_variant(raw[i]));
    }
    Variant result;
    bool call_ok = true;
    String error_msg;
    try {
        const int argc = call_args.size();
        if (argc == 0) result = n->call(method);
        else if (argc == 1) result = n->call(method, call_args[0]);
        else if (argc == 2) result = n->call(method, call_args[0], call_args[1]);
        else if (argc == 3) result = n->call(method, call_args[0], call_args[1], call_args[2]);
        else if (argc == 4) result = n->call(method, call_args[0], call_args[1], call_args[2], call_args[3]);
        else if (argc == 5) result = n->call(method, call_args[0], call_args[1], call_args[2], call_args[3], call_args[4]);
        else return make_error("call_method supports at most 5 arguments");
    } catch (const std::exception &e) {
        call_ok = false;
        error_msg = String("Exception in method '") + method + String("': ") + String(e.what());
    } catch (...) {
        call_ok = false;
        error_msg = String("Unknown exception in method '") + method + String("'");
    }
    if (!call_ok) return make_error(error_msg);
    bool is_nil = result.get_type() == Variant::NIL;
    Dictionary r; r["node_path"] = p; r["method"] = method; r["result"] = variant_to_json(result);
    r["type"] = (int64_t)result.get_type();
    if (is_nil) r["hint"] = "Method returned null. Use @tool script for editor-time custom methods.";
    return r;
}
Dictionary cmd_get_variable(const Dictionary &a) {
    String p = args_string(a, "node_path");
    String var_name = args_string(a, "variable");
    if (var_name.is_empty()) return make_error("missing 'variable'");
    Node *root = get_root(); if (!root) return make_error("no scene");
    Node *n = resolve_node(root, p); if (!n) return make_error("node not found: " + p);
    Variant val = n->get(var_name);
    Dictionary r; r["node_path"] = p; r["variable"] = var_name;
    r["value"] = variant_to_json(val); r["type"] = (int64_t)val.get_type();
    return r;
}
Dictionary cmd_set_variable(const Dictionary &a) {
    String p = args_string(a, "node_path");
    String var_name = args_string(a, "variable");
    if (var_name.is_empty()) return make_error("missing 'variable'");
    if (!a.has("value")) return make_error("missing 'value'");
    Node *root = get_root(); if (!root) return make_error("no scene");
    Node *n = resolve_node(root, p); if (!n) return make_error("node not found: " + p);
    Variant gv = json_to_variant(a["value"]);
    n->set(var_name, gv);
    Variant actual = n->get(var_name);
    Dictionary r; r["node_path"] = p; r["variable"] = var_name;
    r["value"] = variant_to_json(actual); r["success"] = actual.get_type() != Variant::NIL;
    return r;
}
} // namespace

void register_script_helpers(HandlerRegistry &reg) {
    reg.register_tool("call_method", cmd_call_method);
    reg.register_tool("get_variable", cmd_get_variable);
    reg.register_tool("set_variable", cmd_set_variable);
}
} // namespace godot_mcp